Source/JavaScriptCore/runtime/MicrotaskQueue.h

```diff
@@ -33,37 +33,73 @@
 
 namespace JSC {
 
+class JSGlobalObject;
+class MarkedMicrotaskDeque;
+class MicrotaskDispatcher;
 class MicrotaskQueue;
 class VM;
 
 class QueuedTask {
     WTF_MAKE_TZONE_ALLOCATED(QueuedTask);
     friend class MicrotaskQueue;
+    friend class MarkedMicrotaskDeque;
 public:
     static constexpr unsigned maxArguments = 4;
 
-    QueuedTask(MicrotaskIdentifier identifier, JSValue job, JSValue argument0, JSValue argument1, JSValue argument2, JSValue argument3)
-        : m_identifier(identifier)
+    enum class Result : uint8_t {
+        Executed,
+        Discard,
+        Suspended,
+    };
+
+    QueuedTask(RefPtr<MicrotaskDispatcher>&& dispatcher)
+        : m_dispatcher(WTFMove(dispatcher))
+        , m_identifier(MicrotaskIdentifier::generate())
+    {
+    }
+
+    QueuedTask(RefPtr<MicrotaskDispatcher>&& dispatcher, JSGlobalObject* globalObject, JSValue job, JSValue argument0, JSValue argument1, JSValue argument2, JSValue argument3)
+        : m_dispatcher(WTFMove(dispatcher))
+        , m_identifier(MicrotaskIdentifier::generate())
+        , m_globalObject(globalObject)
         , m_job(job)
         , m_arguments { argument0, argument1, argument2, argument3 }
     {
     }
 
-    void run();
+    void setDispatcher(RefPtr<MicrotaskDispatcher>&& dispatcher)
+    {
+        m_dispatcher = WTFMove(dispatcher);
+    }
+
+    Result tryRun();
+    bool isRunnable() const;
 
     MicrotaskIdentifier identifier() const { return m_identifier; }
+    JSGlobalObject* globalObject() const { return m_globalObject; }
+    JSValue job() const { return m_job; }
+    std::span<const JSValue> arguments() const { return std::span { m_arguments }; }
 
 private:
+    RefPtr<MicrotaskDispatcher> m_dispatcher;
```

Comment:
Is there likely going to be many different dispatcher pointers per JSC::VM? I wonder if this could be passed along to the function that runs the microtasks